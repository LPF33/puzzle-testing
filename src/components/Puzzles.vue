<template>
    <div id="puzzles-component">
        <h1>Select a puzzle</h1>
        <div v-for="puzzle in puzzles" :key="puzzle.id" class="puzzle">
            <button @click="selectPuzzle(puzzle)" class="play-button">
                <div>
                    <img :src="require(`../assets/${puzzle.image}`)" />
                    <h2>{{ puzzle.title }}</h2>
                </div>
            </button>
            <Records :puzzleId="puzzle.id" />
        </div>
    </div>
</template>

<script>
import Records from "./Records";
export default {
    name: "Puzzles",
    components: {
        Records,
    },
    data() {
        return {
            puzzles: [
                { id: "owl", image: "owl.jpg", title: "Owl" },
                { id: "eggcozy", image: "eggcozy.jpg", title: "Egg cozy" },
            ],
        };
    },
    methods: {
        selectPuzzle(p) {
            this.$emit("play", p);
        },
    },
};
</script>

<style scoped>
#puzzles-component {
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex-grow: 1;
}

.puzzle {
    position: relative;
    display: flex;
    max-width: 100vw;
    flex-wrap: wrap;
    justify-content: space-evenly;
    align-items: center;
}

.puzzle img {
    width: 300px;
    max-width: 100vw;
    border-radius: 5px;
}

.play-button {
    width: 800px;
    max-width: 100vw;
    height: 100%;
    background-color: #87a7b3;
    cursor: pointer;
    margin: 5px;
    padding: 10px;
    border-radius: 5px;
}

.play-button h2 {
    color: #766161;
    text-shadow: 1px 1px 2px black;
}

h1 {
    color: #766161;
    text-shadow: 1px 1px 2px black;
    background-color: #e1f1dd;
    padding: 10px 0;
    margin-bottom: 10px;
}
</style>
