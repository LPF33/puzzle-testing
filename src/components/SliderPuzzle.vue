<template>
    <h3>"{{ puzzleObj.title }}" puzzle (25 pieces)</h3>
    <h4>Elapsed Time: {{ currentElapsedTime }}</h4>
    <div v-if="checkForWin" class="user-win-game">
        <h3>You have solved the puzzle!!!</h3>
        <button @click="restart">Play again</button>
    </div>
    <div id="puzzle">
        <img
            v-for="(piece, index) of shuffledPuzzleArray"
            :key="index"
            :src="require(`../assets/${puzzleObj.id}/${piece}`)"
            @click="swapImg(index)"
        />
    </div>
    <button @click="closePuzzle" id="close-button">X</button>
</template>

<script>
const correctPuzzleArray = [
    "image_part_001.jpg",
    "image_part_002.jpg",
    "image_part_003.jpg",
    "image_part_004.jpg",
    "image_part_005.jpg",
    "image_part_006.jpg",
    "image_part_007.jpg",
    "image_part_008.jpg",
    "image_part_009.jpg",
    "image_part_010.jpg",
    "image_part_011.jpg",
    "image_part_012.jpg",
    "image_part_013.jpg",
    "image_part_014.jpg",
    "image_part_015.jpg",
    "image_part_016.jpg",
    "image_part_017.jpg",
    "image_part_018.jpg",
    "image_part_019.jpg",
    "image_part_020.jpg",
    "image_part_021.jpg",
    "image_part_022.jpg",
    "image_part_023.jpg",
    "image_part_024.jpg",
    "image_part_025.jpg",
];
export default {
    name: "SliderPuzzle",
    props: ["puzzleObj"],
    data() {
        return {
            correctPuzzleArray,
            shuffledPuzzleArray: [...correctPuzzleArray].sort(
                () => Math.random() - 0.5
            ),
            timerId: null,
            indexedToSwap: [],
            startDateTime: Date.now(),
            currentElapsedTime: null,
        };
    },
    mounted() {
        this.timerId = setInterval(this.elapsedTime, 1000);
    },
    beforeUnmount() {
        console.log("destroyed");
        clearInterval(this.timerId);
    },
    emits: ["close"],
    methods: {
        closePuzzle() {
            clearInterval(this.timerId);
            this.$emit("close");
        },
        elapsedTime() {
            const elpased = new Date(Date.now() - this.startDateTime);
            this.currentElapsedTime = elpased.toISOString().slice(14, 19);
        },
        swapImg(index) {
            if (this.indexedToSwap.length < 2) {
                this.indexedToSwap.push(index);
            }

            if (this.indexedToSwap.length === 2) {
                const [first, second] = this.indexedToSwap;
                [
                    this.shuffledPuzzleArray[first],
                    this.shuffledPuzzleArray[second],
                ] = [
                    this.shuffledPuzzleArray[second],
                    this.shuffledPuzzleArray[first],
                ];

                this.indexedToSwap = [];
            }
        },
        saveRecord() {
            try {
                const records =
                    JSON.parse(localStorage.getItem("records")) || [];
                records.push([
                    this.puzzleObj.id,
                    Date.now() - this.startDateTime,
                ]);
                localStorage.setItem("records", JSON.stringify(records));
            } catch (error) {
                console.log("could not save record");
            }
        },
        restart() {
            this.shuffledPuzzleArray = [...correctPuzzleArray].sort(
                () => Math.random() - 0.5
            );
            this.startDateTime = Date.now();
            this.timerId = setInterval(this.elapsedTime, 1000);
        },
    },
    computed: {
        checkForWin() {
            for (let i = 0; i < correctPuzzleArray.length; i++) {
                if (correctPuzzleArray[i] !== this.shuffledPuzzleArray[i]) {
                    return false;
                }
            }
            clearInterval(this.timerId);
            this.saveRecord();
            return true;
        },
    },
};
</script>
<style scoped>
#puzzle {
    display: grid;
    grid-template: repeat(5, minmax(50px, 100px)) / repeat(
            5,
            minmax(50px, 100px)
        );
    align-items: center;
    justify-content: center;
    margin-bottom: 10px;
}

#puzzle img {
    width: 100%;
    height: 100%;
}

#close-button {
    position: fixed;
    top: 10px;
    right: 10px;
    padding: 10px;
    border-radius: 5px;
    background-color: #766161;
    color: white;
    font-weight: bold;
}

h3 {
    color: #766161;
    text-shadow: 1px 1px 2px black;
    background-color: #e1f1dd;
    padding: 10px 0;
}

h4 {
    color: black;
    text-shadow: 1px 1px 2px black;
    padding: 10px 0;
}
</style>
