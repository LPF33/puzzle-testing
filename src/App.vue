<template>
    <div id="app">
        <h1>Welcome to the PUZZLE world!</h1>
        <Puzzles v-if="!selectedPuzzle" v-on:play="play" />
        <SliderPuzzle
            :puzzleObj="selectedPuzzle"
            v-if="selectedPuzzle"
            @close="setNull"
        />
    </div>
</template>

<script>
import Puzzles from "./components/Puzzles";
import SliderPuzzle from "./components/SliderPuzzle";

export default {
    name: "App",
    components: {
        Puzzles,
        SliderPuzzle,
    },
    data() {
        return {
            selectedPuzzle: null,
        };
    },
    methods: {
        play(puzzle) {
            this.selectedPuzzle = puzzle;
        },
        setNull() {
            this.selectedPuzzle = null;
        },
    },
};
</script>

<style>
* {
    margin: 0;
    border: 0;
    padding: 0;
    box-sizing: border-box;
    text-align: center;
}
#app {
    background-color: #cdc7be;
    width: 100%;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
}

#app > h1 {
    color: #766161;
    text-shadow: 1px 1px 2px black;
    background-color: #e1f1dd;
    padding: 10px 0;
}
</style>
