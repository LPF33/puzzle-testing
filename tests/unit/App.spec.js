import { mount, shallowMount } from "@vue/test-utils";
import App from "@/App";
import Puzzles from "@/components/Puzzles.vue";
import SliderPuzzle from "@/components/SliderPuzzle.vue";

test("App is mounted correctly when initially selectedPuzzle is null", () => {
    const wrapper = mount(App);
    expect(wrapper.findAllComponents(Puzzles)).toHaveLength(1);
    expect(wrapper.findAllComponents(SliderPuzzle)).toHaveLength(0);
});

test("App is mounted correctly when selectedPuzzle is not null", () => {
    const wrapper = shallowMount(App, {
        data() {
            return {
                selectedPuzzle: {},
            };
        },
    });
    expect(wrapper.findAllComponents(Puzzles)).toHaveLength(0);
    expect(wrapper.findAllComponents(SliderPuzzle)).toHaveLength(1);
});

test("When play method is envoked show SliderPuzzle and unmount Puzzles component", async () => {
    const wrapper = shallowMount(App);
    wrapper.vm.play({});
    await wrapper.vm.$nextTick();
    expect(wrapper.findAllComponents(Puzzles)).toHaveLength(0);
    expect(wrapper.findAllComponents(SliderPuzzle)).toHaveLength(1);
});

test("When setNull method is called, Puzzles component gets rendered and SliderPuzzle gets unmounted", async () => {
    const wrapper = shallowMount(App);
    wrapper.vm.play({});
    await wrapper.vm.$nextTick();
    expect(wrapper.findAllComponents(Puzzles)).toHaveLength(0);
    expect(wrapper.findAllComponents(SliderPuzzle)).toHaveLength(1);
    wrapper.vm.setNull();
    await wrapper.vm.$nextTick();
    expect(wrapper.findAllComponents(Puzzles)).toHaveLength(1);
    expect(wrapper.findAllComponents(SliderPuzzle)).toHaveLength(0);
});

test("When Puzzle Owl gets mounted and triggers close -> SliderPuzzle gets unmounted", async () => {
    const wrapper = shallowMount(App);

    wrapper.vm.play({ id: "owl", image: "owl.jpg", title: "Owl" });
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();

    wrapper.findComponent(SliderPuzzle).trigger("close");
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.findAllComponents(Puzzles)).toHaveLength(1);
    expect(wrapper.findAllComponents(SliderPuzzle)).toHaveLength(0);
});
