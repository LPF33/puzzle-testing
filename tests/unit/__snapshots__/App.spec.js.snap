// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When Puzzle Owl gets mounted and triggers close -> SliderPuzzle gets unmounted 1`] = `
<div id="app">
  <h1>Welcome to the PUZZLE world!</h1>
  <!--v-if-->
  <slider-puzzle-stub puzzleobj="[object Object]"></slider-puzzle-stub>
</div>
`;

exports[`When Puzzle Owl gets mounted and triggers close -> SliderPuzzle gets unmounted 2`] = `
<div id="app">
  <h1>Welcome to the PUZZLE world!</h1>
  <puzzles-stub></puzzles-stub>
  <!--v-if-->
</div>
`;
